modules/secp256k1/module.cpp

```diff
@@ -0,0 +1,160 @@
+#define template c_template // avoid C++ keyword conflict just during includes
+
+extern "C"
+{
+#include <secp256k1.h>
+#include <secp256k1_ecdh.h>
+}
+
+#undef template
+
+#include <iomanip>
+#include <sstream>
+#include <assert.h>
+#include "module.h"
+
+#define SECP256K1_PUBKEY_COMPRESSED_LEN 33
+#define SECP256K1_SIGNATURE_COMPACT_LEN 64
+#define SECP256K1_SIGNATURE_DER_MAX_LEN 74
+#define SECP256K1_SHARED_SECRET_LEN 32
+
+secp256k1_context *secp256k1_ctx;
+
+void init(int *argc, char ***argv)
+{
+	if (!secp256k1_ctx)
+	{
+		secp256k1_ctx = secp256k1_context_create(SECP256K1_CONTEXT_NONE);
+		assert(secp256k1_ctx);
+	}
+}
+
+std::string hex_encode(const unsigned char *data, size_t len)
+{
+	std::ostringstream oss;
+	oss << std::hex << std::setfill('0');
+	for (size_t i = 0; i < len; ++i)
+	{
+		oss << std::setw(2) << static_cast<int>(data[i]);
+	}
+	return oss.str();
+}
+
+int create_pubkey_from_privkey(const uint8_t *privkey, secp256k1_pubkey &pubkey)
+{
+	int ret = 0;
+
+	ret = secp256k1_ec_seckey_verify(secp256k1_ctx, privkey);
+	ret = ret && secp256k1_ec_pubkey_create(secp256k1_ctx, &pubkey, privkey);
+
+	return ret;
+}
+
+int create_signature_from_hash(const uint8_t *privkey, const uint8_t *hash, secp256k1_ecdsa_signature &signature)
+{
+	int ret = 0;
+
+	ret = secp256k1_ec_seckey_verify(secp256k1_ctx, privkey);
+	ret = ret && secp256k1_ecdsa_sign(secp256k1_ctx, &signature, hash, privkey, nullptr, nullptr);
+	ret = ret && secp256k1_ecdsa_signature_normalize(secp256k1_ctx, &signature, &signature);
+
+	return ret;
+}
+
+std::optional<std::string> secp256k1_private_to_public(std::span<const uint8_t> buffer)
+{
+	int ret = 0;
+	secp256k1_pubkey pubkey;
+	size_t pubkey_len = SECP256K1_PUBKEY_COMPRESSED_LEN;
+	std::vector<uint8_t> pubkey_compressed(pubkey_len);
+
+	ret = create_pubkey_from_privkey(buffer.data(), pubkey);
+	ret = ret && secp256k1_ec_pubkey_serialize(secp256k1_ctx, pubkey_compressed.data(), &pubkey_len, &pubkey, SECP256K1_EC_COMPRESSED);
+	if (!ret)
+	{
+		return std::nullopt;
+	}
+
+	return hex_encode(pubkey_compressed.data(), pubkey_len);
+}
```

Comment:
This aligns with what Bruno commented. We should only skip when parsing the private key since decred secp256k1 is much more lenient. In this context, if secp256k1 can parse the private key, then it should be able to create a public key from it. If not, we will be able to catch this bug by returning an empty string. We should only return a null pointer in very specific cases where we want to skip the string comparison in driver.cpp.

```suggestion
std::optional<std::string> secp256k1_private_to_public(std::span<const uint8_t> buffer)
{
	secp256k1_pubkey pubkey;
	size_t pubkey_len = SECP256K1_PUBKEY_COMPRESSED_LEN;
	std::vector<uint8_t> pubkey_compressed(pubkey_len);
	const uint8_t * privkey = buffer.data();

	if (!secp256k1_ec_seckey_verify(secp256k1_ctx, privkey)) {
		return std::nullopt;
	}
	if (!secp256k1_ec_pubkey_create(secp256k1_ctx, &pubkey, privkey)) {
		return "";
	}

	if (!secp256k1_ec_pubkey_serialize(secp256k1_ctx, pubkey_compressed.data(), &pubkey_len, &pubkey, SECP256K1_EC_COMPRESSED)) {
		return "";
	}

	return hex_encode(pubkey_compressed.data(), pubkey_len);
}
```